packages/data/src/world/mq/ice_cavern_mq.yml

```diff
@@ -2,42 +2,98 @@
+    "Ice Cavern Main Room": "true"
   locations:
     "MQ Ice Cavern Pot First Room 1": "true"
     "MQ Ice Cavern Pot First Room 2": "true"
+"Ice Cavern Main Room":
+  dungeon: IC
+  exits:
+    "Ice Cavern After Initial Ice Block": "true"
+    "Ice Cavern Map Room": "event(ICE_MQ_MAIN_ENEMIES_CLEAR)"
+    "Ice Cavern Compass Room": "(has_blue_fire && is_adult) || climb_anywhere || hookshot_anywhere"
+    "Ice Cavern Vanilla Push Block Room": "has_blue_fire || climb_anywhere || hookshot_anywhere"
+  events:
+    ICE_MQ_MAIN_ENEMIES_CLEAR: "soul_wolfos && soul_freezard && (has_weapon || can_use_sticks || has_explosives)"
+  locations:
     "MQ Ice Cavern Pot Main Room 1": "true"
     "MQ Ice Cavern Pot Main Room 2": "true"
     "MQ Ice Cavern Pot Main Room 3": "true"
     "MQ Ice Cavern Pot Main Room 4": "true"
 "Ice Cavern Map Room":
   dungeon: IC
+  exits:
+    "Ice Cavern Main Room": "true"
   events:
-    BLUE_FIRE: "has_bottle"
+    BLUE_FIRE: "has_bottle && event(ICE_MQ_MAP_STALAGMITES)"
+    ICE_MQ_MAP_STALAGMITES: "true" #Need to get past ice stalagmites to reach Blue Fire, trivial until action shuffle is implemented as we can roll between them (albeit finnickly).
+    ICE_MQ_MAP_SWITCH: "has_weapon || has_bombs || can_use_mask_blast || has_bombchu || can_use_bow || can_use_slingshot || can_hookshot || can_use_sticks || can_hammer || can_boomerang" #Many ways to hit crystal switch
   locations:
-    "MQ Ice Cavern Map Chest": "has_blue_fire"
+    "MQ Ice Cavern Map Chest": "event(ICE_MQ_MAP_SWITCH) && has_blue_fire"
 "Ice Cavern Compass Room":
   dungeon: IC
+  exits:
+    "Ice Cavern Main Room": "has_blue_fire || climb_anywhere || hookshot_anywhere" #Climbing works if you start further down the hall
+  events:
+    BLUE_FIRE: "has_bottle"
   locations:
     "MQ Ice Cavern Compass Chest": "true"
-    "MQ Ice Cavern Piece of Heart": "has_explosives"
-    "MQ Ice Cavern GS Compass Room": "gs && (can_play_time || has_blue_fire_arrows || can_boomerang)"
+    "MQ Ice Cavern Piece of Heart": "has_explosives || (can_use_mask_blast && (is_child || (is_adult && has_shield)))" #Blast Mask can only hit when adult if crouching with shield.
+    "MQ Ice Cavern GS Compass Room": "gs && ((can_play_time && has_blue_fire && can_damage_skull) || has_blue_fire_arrows || can_boomerang)"
     "MQ Ice Cavern Pot Compass Room 1": "true"
     "MQ Ice Cavern Pot Compass Room 2": "true"
-"Ice Cavern Big Room":
+"Ice Cavern Vanilla Push Block Room":
+  dungeon: IC
+  exits:
+    "Ice Cavern Main Room": "has_blue_fire || hookshot_anywhere"
+    "Ice Cavern Vanilla Push Block Room Scarecrow Ledge": "climb_anywhere || hookshot_anywhere || scarecrow_hookshot ||(is_adult && trick(OOT_MQ_ICE_SCARE_NOTHING)) || (has_hover_boots && (can_longshot || (is_adult && trick(OOT_MQ_ICE_SCARE_HOVER))))"
+    "Ice Cavern Past Red Ice Before Shiek": "climb_anywhere || hookshot_anywhere || (has_blue_fire && (is_adult || (can_play_time && has_hover_boots)))"
```

Comment:
Couldn't this just be Before Shiek if there isn't one for not being past the red ice?

Reply:
This is in prep for Door Rando.

Passed Red Ice Before Shielk > Serenade Entry Door > Serenade Exit Door > Serenade Room

Reply:
After some thinking, I recognize specifying being past the red ice is necessary, due to vanilla Ice Cavern